### PR TITLE
Add explicit aligned and unaligned read benches

### PR DESCRIPTION
**Pull request by sffc**
I am adding another benchmark to illustrate the performance of aligned and unaligned reads.

I ran these benchmarks 5 times in a row on my x86_64 laptop.  Results:

<details>
Run 1:

```
zerovec/sum/sample/slice                                                                             
                        time:   [5.7847 ns 5.8072 ns 5.8329 ns]

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.3018 ns 5.3197 ns 5.3393 ns]
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.1618 ns 5.1976 ns 5.2479 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
```

Run 2:

```
zerovec/sum/sample/slice                                                                             
                        time:   [5.2745 ns 5.2897 ns 5.3091 ns]
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) low mild
  7 (7.00%) high mild
  3 (3.00%) high severe

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.2183 ns 5.2403 ns 5.2657 ns]
Found 7 outliers among 100 measurements (7.00%)
  6 (6.00%) high mild
  1 (1.00%) high severe

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.2843 ns 5.3101 ns 5.3400 ns]
Found 7 outliers among 100 measurements (7.00%)
  6 (6.00%) high mild
  1 (1.00%) high severe
```

Run 3:

```
zerovec/sum/sample/slice                                                                             
                        time:   [5.3272 ns 5.3501 ns 5.3773 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.3391 ns 5.3758 ns 5.4189 ns]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.3387 ns 5.3549 ns 5.3714 ns]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
```

Run 4:

```
zerovec/sum/sample/slice                                                                             
                        time:   [4.9303 ns 5.0362 ns 5.1684 ns]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.2726 ns 5.2883 ns 5.3053 ns]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.4446 ns 5.4776 ns 5.5203 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

Run 5:

```
zerovec/sum/sample/slice                                                                             
                        time:   [5.3702 ns 5.4031 ns 5.4386 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.2765 ns 5.3041 ns 5.3413 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.2893 ns 5.3111 ns 5.3399 ns]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high severe
```
</details>

In tabular form, if I take the median of the 5 runs:

| Read Type | Memory Type | 95% Confidence Interval |
|---|---|---|
| Aligned | Aligned | 5.3272 < 5.3501 < 5.3773 |
| Unaligned | Aligned | 5.2765 < 5.3041 < 5.3393 |
| Unaligned | Unaligned | 5.2893 < 5.3111 < 5.3400 |

In words: the confidence intervals for all 3 types are overlapping, and unaligned reads do not show any regression over aligned reads, especially if they come from aligned memory.

### CODE REVIEW

**Comment by Manishearth** [comment]
```
Testing zerovec/sum/sample/zerovec_unaligned
thread 'main' panicked at 'attempt to add with overflow', /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/iter/traits/accum.rs:141:1
stack backtrace:
```

what

File: utils/zerovec/benches/zerovec.rs
Comment (Manishearth) [inline utils/zerovec/benches/zerovec.rs:82]: praise: clever way to avoid having to introduce more samples

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
Rewrote with `wrapping_add`.

Same type of figures:

```
zerovec/sum/sample/slice                                                                             
                        time:   [5.0883 ns 5.0957 ns 5.1056 ns]
                        change: [-5.7891% -5.2323% -4.6691%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  8 (8.00%) high mild
  4 (4.00%) high severe

zerovec/sum/sample/zerovec_aligned                                                                             
                        time:   [5.1042 ns 5.1142 ns 5.1252 ns]
                        change: [-4.0638% -3.4758% -2.9220%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe

zerovec/sum/sample/zerovec_unaligned                                                                             
                        time:   [5.1047 ns 5.1144 ns 5.1258 ns]
                        change: [-6.3770% -4.5432% -3.1792%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe
```

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
Waiting on review from @hsivonen

**Review by hsivonen — APPROVED** [review]

**Comment by sffc** [comment]
I recently obtained access to an M2 MacBook Air, so I ran the benches on this architecture.

```
zerovec/sum/sample/slice
                        time:   [2.6550 ns 2.6555 ns 2.6561 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

zerovec/sum/sample/zerovec_aligned
                        time:   [4.1294 ns 4.1310 ns 4.1324 ns]
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe

zerovec/sum/sample/zerovec_unaligned
                        time:   [4.1282 ns 4.1301 ns 4.1317 ns]
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe

zerovec/binary_search/sample/slice
                        time:   [4.3084 ns 4.3197 ns 4.3314 ns]
Found 15 outliers among 100 measurements (15.00%)
  12 (12.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high severe

zerovec/binary_search/sample/zerovec
                        time:   [4.3071 ns 4.3082 ns 4.3095 ns]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe
```

For the read-heavy summation operation, on aarch64, there is a boost for a fully aligned slice, moreso than on x86_64.

For the binary search operation, there is no difference.

Here is the aarch64 assembly code for the sum operation, first the slice version then the zerovec version:

```
00000001000088a8 <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E>:
1000088a8: fd 7b bf a9 	stp	x29, x30, [sp, #-16]!
1000088ac: fd 03 00 91 	mov	x29, sp
1000088b0: 01 08 01 8b 	add	x1, x0, x1, lsl #2
1000088b4: 16 8c 05 94 	bl	0x10016b90c <__ZN4core4iter8adapters6copied15Copied$LT$I$GT$3new17h0f63329ad409c938E>
1000088b8: 1f 00 01 eb 	cmp	x0, x1
1000088bc: e0 00 00 54 	b.eq	0x1000088d8 <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x30>
1000088c0: 28 00 00 cb 	sub	x8, x1, x0
1000088c4: 08 11 00 d1 	sub	x8, x8, #4
1000088c8: 1f 71 00 f1 	cmp	x8, #28
1000088cc: a2 00 00 54 	b.hs	0x1000088e0 <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x38>
1000088d0: 08 00 80 52 	mov	w8, #0
1000088d4: 16 00 00 14 	b	0x10000892c <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x84>
1000088d8: 08 00 80 52 	mov	w8, #0
1000088dc: 18 00 00 14 	b	0x10000893c <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x94>
1000088e0: 08 fd 42 d3 	lsr	x8, x8, #2
1000088e4: 0a 05 00 91 	add	x10, x8, #1
1000088e8: 49 ed 7d 92 	and	x9, x10, #0x7ffffffffffffff8
1000088ec: 08 40 00 91 	add	x8, x0, #16
1000088f0: 00 e4 00 6f 	movi.2d	v0, #0000000000000000
1000088f4: eb 03 09 aa 	mov	x11, x9
1000088f8: 01 e4 00 6f 	movi.2d	v1, #0000000000000000
1000088fc: 02 8d 7f ad 	ldp	q2, q3, [x8, #-16]
100008900: 40 84 a0 4e 	add.4s	v0, v2, v0
100008904: 61 84 a1 4e 	add.4s	v1, v3, v1
100008908: 08 81 00 91 	add	x8, x8, #32
10000890c: 6b 21 00 f1 	subs	x11, x11, #8
100008910: 61 ff ff 54 	b.ne	0x1000088fc <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x54>
100008914: 20 84 a0 4e 	add.4s	v0, v1, v0
100008918: 00 b8 b1 4e 	addv.4s	s0, v0
10000891c: 08 00 26 1e 	fmov	w8, s0
100008920: 5f 01 09 eb 	cmp	x10, x9
100008924: c0 00 00 54 	b.eq	0x10000893c <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x94>
100008928: 00 08 09 8b 	add	x0, x0, x9, lsl #2
10000892c: 09 44 40 b8 	ldr	w9, [x0], #4
100008930: 28 01 08 0b 	add	w8, w9, w8
100008934: 1f 00 01 eb 	cmp	x0, x1
100008938: a1 ff ff 54 	b.ne	0x10000892c <__ZN7zerovec24zerovec_sum_sample_slice17h8e3943df1b4541b0E+0x84>
10000893c: e0 03 08 aa 	mov	x0, x8
100008940: fd 7b c1 a8 	ldp	x29, x30, [sp], #16
100008944: c0 03 5f d6 	ret

0000000100008948 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E>:
100008948: fd 7b bf a9 	stp	x29, x30, [sp, #-16]!
10000894c: fd 03 00 91 	mov	x29, sp
100008950: 01 08 01 8b 	add	x1, x0, x1, lsl #2
100008954: ee 8b 05 94 	bl	0x10016b90c <__ZN4core4iter8adapters6copied15Copied$LT$I$GT$3new17h0f63329ad409c938E>
100008958: f4 bd 05 94 	bl	0x100178128 <__ZN4core4iter8adapters3map16Map$LT$I$C$F$GT$3new17h09ff1c9317dc4631E>
10000895c: 1f 00 01 eb 	cmp	x0, x1
100008960: e0 00 00 54 	b.eq	0x10000897c <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x34>
100008964: 28 00 00 cb 	sub	x8, x1, x0
100008968: 08 11 00 d1 	sub	x8, x8, #4
10000896c: 1f 71 00 f1 	cmp	x8, #28
100008970: a2 00 00 54 	b.hs	0x100008984 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x3c>
100008974: 08 00 80 52 	mov	w8, #0
100008978: 17 00 00 14 	b	0x1000089d4 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x8c>
10000897c: 08 00 80 52 	mov	w8, #0
100008980: 19 00 00 14 	b	0x1000089e4 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x9c>
100008984: 08 fd 42 d3 	lsr	x8, x8, #2
100008988: 09 05 00 91 	add	x9, x8, #1
10000898c: 2b ed 7d 92 	and	x11, x9, #0x7ffffffffffffff8
100008990: 6a f5 7e d3 	lsl	x10, x11, #2
100008994: 08 40 00 91 	add	x8, x0, #16
100008998: 00 e4 00 6f 	movi.2d	v0, #0000000000000000
10000899c: ec 03 0b aa 	mov	x12, x11
1000089a0: 01 e4 00 6f 	movi.2d	v1, #0000000000000000
1000089a4: 02 8d 7f ad 	ldp	q2, q3, [x8, #-16]
1000089a8: 40 84 a0 4e 	add.4s	v0, v2, v0
1000089ac: 61 84 a1 4e 	add.4s	v1, v3, v1
1000089b0: 08 81 00 91 	add	x8, x8, #32
1000089b4: 8c 21 00 f1 	subs	x12, x12, #8
1000089b8: 61 ff ff 54 	b.ne	0x1000089a4 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x5c>
1000089bc: 20 84 a0 4e 	add.4s	v0, v1, v0
1000089c0: 00 b8 b1 4e 	addv.4s	s0, v0
1000089c4: 08 00 26 1e 	fmov	w8, s0
1000089c8: 3f 01 0b eb 	cmp	x9, x11
1000089cc: c0 00 00 54 	b.eq	0x1000089e4 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x9c>
1000089d0: 00 00 0a 8b 	add	x0, x0, x10
1000089d4: 09 44 40 b8 	ldr	w9, [x0], #4
1000089d8: 28 01 08 0b 	add	w8, w9, w8
1000089dc: 1f 00 01 eb 	cmp	x0, x1
1000089e0: a1 ff ff 54 	b.ne	0x1000089d4 <__ZN7zerovec26zerovec_sum_sample_zerovec17hd8e97ffb07cd3880E+0x8c>
1000089e4: e0 03 08 aa 	mov	x0, x8
1000089e8: fd 7b c1 a8 	ldp	x29, x30, [sp], #16
1000089ec: c0 03 5f d6 	ret
```

**Comment by sffc** [comment]
The only differences between the two functions I can see:

1. The zerovec version invokes an extra call to `core::iter::adapters::map::Map::new`, perhaps originating in the `.iter()` function that calls `.map(T::from_unaligned)`
2. The slice version appears to put `lsl` into the `add` operation, but the zerovec version has it as a separate operation

I can't find any more differences, other than addresses and register names being different.

On x86 I usually see load operations annotated with an "unaligned" bit. But here, the only load operations appear to be `ldp` and `ldr`, and there's no distinction about them being aligned vs unaligned.

So I really don't know where the perf difference comes from.

Here is the code, by the way:

```rust
#[cfg(feature = "bench")]
#[inline(never)]
fn zerovec_sum_sample_slice(slice: &[u32]) -> u32 {
    slice
        .iter()
        .copied()
        .fold(0u32, |sum, val| sum.wrapping_add(val))
}

#[cfg(feature = "bench")]
#[inline(never)]
fn zerovec_sum_sample_zerovec(v: &[zerovec::ule::RawBytesULE<4>]) -> u32 {
    ZeroSlice::<u32>::from_ule_slice    (v)
        .iter()
        .fold(0u32, |sum, val| sum.wrapping_add(val))
}
```

**Pull request merged**
